### PR TITLE
a small handful of optimizations for 2.12

### PR DESCRIPTION
**Pull request by hrhino**
Benchmarking results are [here](https://gist.github.com/hrhino/556f3c080160fde1473338d25c8d238c). I removed some commits which turned out to be counterproductive, so the hashes are wrong; the commit messages are correct, though. I'll run some more over the weekend.

cc @mkeskells @rorygraves

### CODE REVIEW

File: src/reflect/scala/reflect/internal/Trees.scala [hunk 1]
```diff
@@ -205,11 +205,8 @@ trait Trees extends api.Trees {
       new ForeachPartialTreeTraverser(pf).traverse(this)
     }
 
-    def changeOwner(pairs: (Symbol, Symbol)*): Tree = {
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/Trees.scala]: Let's leave this version around as an overload, it is likely to be called from compiler plugins.

File: src/reflect/scala/reflect/internal/Symbols.scala [hunk 2]
```diff
@@ -3594,6 +3594,14 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
...
   lazy val NoSymbol: NoSymbol = makeNoSymbol
 
+  /** A symbol representing a language feature, along with some useful information. */
+  case class LanguageFeature(sym: Symbol, name: String) {
+    def tpe = sym.tpe
+    val AnnotationInfo(_, List(
+          Literal(Constant(desc: String)),
+          Literal(Constant(required: Boolean))
+        ), _) = sym.getAnnotation(LanguageFeatureAnnot).get
+  }
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/Symbols.scala]: During the `scala` corpus [benchmark](https://scala-ci.typesafe.com/view/scala-bench/job/compiler-benchmark/1813/console), which has the language features defined in sources:

```
[error] 	Suppressed: java.util.NoSuchElementException: None.get
[error] 		at scala.None$.get(Option.scala:349)
[error] 		at scala.None$.get(Option.scala:347)
[error] 		at scala.reflect.internal.Symbols$LanguageFeature.<init>(Symbols.scala:3603)
[error] 		at scala.reflect.internal.Definitions$DefinitionsClass.getLanguageFeature(Definitions.scala:1222)
[error] 		at scala.reflect.internal.Definitions$DefinitionsClass$RunDefinitions.ImplicitConversionsFeature$lzycompute(Definitions.scala:1539)
[error] 		at scala.reflect.internal.Definitions$DefinitionsClass$RunDefinitions.ImplicitConversionsFeature(Definitions.scala:1539)
[error] 		at scala.tools.nsc.typechecker.Typers$Typer.typedDefDef(Typers.scala:2334)
[error] 		at scala.tools.nsc.typechecker.Typers$Typer.typedMemberDef$1(Typers.scala:5545)
[error] 		at scala.tools.nsc.typechecker.Typers$Typer.typed1(Typers.scala:5596)
```

Comment (hrhino) [reply src/reflect/scala/reflect/internal/Symbols.scala]: ... yikes. It bootstrapped just fine (obviously). I'll take a look as to why, or just get rid of the commit if it's not worth it.

Comment (retronym) [reply src/reflect/scala/reflect/internal/Symbols.scala]: The point of difference is that over in compiler-benchmark we jointly compile reflect/library/compiler sources, which seems to extend some weird corner cases with bootstrapping

**Comment by hrhino** [comment]
Will come back to this today.

**Comment by adriaanm** [comment]
@hrhino ping

**Comment by hrhino** [comment]
one of these is not pulling its weight. looking into it, then I'll rebase.

(sorry for the extended absence, I've been really busy sorting out job stuff.)

**Comment by adriaanm** [comment]
I was passing by, so decided to take the tedium out of it for you. Hope you don't mind my push -f (from 9614ee2 to a227295). I restored the changeOwner overload, and moved the dubious commit last.

**Comment by hrhino** [comment]
Thanks, @adriaanm.

I removed the dubious commit (although I still can't get it to fail on the benchmark) as not worth the effort. I reran [the benchmarks](https://gist.github.com/hrhino/2cfa7da77655042350025e2adab34cb8) which show that b2c493d525f476d46095413832ad04a713c12bbc is most of the improvement. 

I'm gonna take the WIP label off. Let me know what you think.

**Review by retronym — APPROVED** [review]

**Pull request merged**